File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConformance/EventsRunner.swift
import Foundation
import IAIsoAudit
import IAIsoCore

/// Runs the events section of the conformance suite.
public enum EventsRunner {
    public static let TOLERANCE: Double = 1e-9

    public static func run(specRoot: String) throws -> [VectorResult] {
        let path = "\(specRoot)/events/vectors.json"
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        guard let doc = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let vectors = doc["vectors"] as? [[String: Any]] else {
            return [VectorResult.fail("events", "load", "could not parse vectors.json")]
        }
        return vectors.map(runOne)
    }

    private static func runOne(_ v: [String: Any]) -> VectorResult {
        let name = (v["name"] as? String) ?? "?"
        var cfg = PressureConfig.defaults
        if let c = v["config"] as? [String: Any] {
            if let n = num(c["escalation_threshold"]) { cfg.escalationThreshold = n }
            if let n = num(c["release_threshold"]) { cfg.releaseThreshold = n }
            if let n = num(c["dissipation_per_step"]) { cfg.dissipationPerStep = n }
            if let n = num(c["dissipation_per_second"]) { cfg.dissipationPerSecond = n }
            if let n = num(c["token_coefficient"]) { cfg.tokenCoefficient = n }
            if let n = num(c["tool_coefficient"]) { cfg.toolCoefficient = n }
            if let n = num(c["depth_coefficient"]) { cfg.depthCoefficient = n }
            if let b = c["post_release_lock"] as? Bool { cfg.postReleaseLock = b }
        }
        let clockSeq: [Double] = (v["clock"] as? [Any])?.compactMap { num($0) } ?? [0.0]
        let clk = ScriptedClock(clockSeq)

        let sink = MemorySink()
        let execId = (v["execution_id"] as? String) ?? "?"
        let engine: PressureEngine
        do {
            engine = try PressureEngine(
                config: cfg,
                options: EngineOptions(executionId: execId,
                                       auditSink: sink, clock: clk, timestampClock: clk))
        } catch {
            return VectorResult.fail("events", name, "engine init failed: \(error)")
        }

        let resetAfterStep = v["reset_after_step"] as? Int
        let steps = (v["steps"] as? [[String: Any]]) ?? []
        for (i, step) in steps.enumerated() {
            if let r = step["reset"] as? Bool, r {
                engine.reset()
            } else {
                var work = StepInput()
                if let n = step["tokens"] as? Int { work.tokens = Int64(n) }
                if let n = step["tool_calls"] as? Int { work.toolCalls = Int64(n) }
                if let n = step["depth"] as? Int { work.depth = Int64(n) }
                if let s = step["tag"] as? String { work.tag = s }
                engine.step(work)
            }
            if let after = resetAfterStep, (i + 1) == after {
                engine.reset()
            }
        }

        let got = sink.events
        guard let expected = v["expected_events"] as? [[String: Any]] else {
            return VectorResult.fail("events", name, "missing expected_events")
        }
        if got.count != expected.count {
            return VectorResult.fail("events", name,
                "event count: got \(got.count), want \(expected.count)")
        }
        for (i, exp) in expected.enumerated() {
            let actual = got[i]
            if let v = exp["schema_version"] as? String, !v.isEmpty,
               actual.schemaVersion != v {
                return VectorResult.fail("events", name,
                    "event \(i) schema_version: got \(actual.schemaVersion), want \(v)")
            }
            if let v = exp["execution_id"] as? String, !v.isEmpty,
               actual.executionId != v {
                return VectorResult.fail("events", name,
                    "event \(i) execution_id: got \(actual.executionId), want \(v)")
            }
            if let k = exp["kind"] as? String, actual.kind != k {
                return VectorResult.fail("events", name,
                    "event \(i) kind: got \(actual.kind), want \(k)")
            }
            if let dataExp = exp["data"] as? [String: Any] {
                if !dataMatches(actual: actual.data, want: dataExp) {
                    return VectorResult.fail("events", name,
                        "event \(i) data mismatch")
                }
            }
        }
        return VectorResult.pass("events", name)
    }

    private static func dataMatches(actual: [String: AnyJSON], want: [String: Any]) -> Bool {
        for (k, w) in want {
            let got = actual[k]
            if !valueEqual(got: got, want: w) { return false }
        }
        return true
    }

    private static func valueEqual(got: AnyJSON?, want: Any) -> Bool {
        if want is NSNull {
            if got == nil { return true }
            if case .null = got! { return true }
            return false
        }
        guard let got = got else { return false }
        // Distinguish Bool from numeric on Apple platforms where
        // JSONSerialization returns NSNumber for both. Without this check,
        // `1` (NSNumber) would match `Bool` first and incorrectly equal `true`.
        if let n = want as? NSNumber {
            #if canImport(ObjectiveC)
            if String(cString: n.objCType) == "c" {
                if case .bool(let g) = got { return g == n.boolValue }
                return false
            }
            #else
            // On Linux Foundation, Bool is its own type, not bridged.
            if let b = want as? Bool, type(of: want) == Bool.self {
                if case .bool(let g) = got { return g == b }
                return false
            }
            #endif
            // Numeric NSNumber.
            let nd = n.doubleValue
            if case .int(let i) = got { return abs(Double(i) - nd) <= TOLERANCE }
            if case .double(let d) = got { return abs(d - nd) <= TOLERANCE }
            return false
        }
        if let b = want as? Bool {
            if case .bool(let g) = got { return g == b }
            return false
        }
        if let n = num(want) {
            if case .int(let i) = got { return abs(Double(i) - n) <= TOLERANCE }
            if case .double(let d) = got { return abs(d - n) <= TOLERANCE }
            return false
        }
        if let s = want as? String {
            if case .string(let g) = got { return g == s }
            return false
        }
        if let arr = want as? [Any] {
            guard case .array(let g) = got, g.count == arr.count else { return false }
            for (i, w) in arr.enumerated() {
                if !valueEqual(got: g[i], want: w) { return false }
            }
            return true
        }
        if let m = want as? [String: Any] {
            guard case .object(let g) = got else { return false }
            for (k, w) in m {
                if !valueEqual(got: g[k], want: w) { return false }
            }
            return true
        }
        return false
    }

    private static func num(_ v: Any?) -> Double? {
        if let n = v as? NSNumber {
            #if canImport(ObjectiveC)
            if String(cString: n.objCType) == "c" { return nil }
            #endif
            return n.doubleValue
        }
        if let d = v as? Double { return d }
        if let i = v as? Int { return Double(i) }
        if let i = v as? Int64 { return Double(i) }
        return nil
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/conformance/src/events_runner.rs
use crate::{config_from_value, scripted_clock, VectorResult};
use iaiso_audit::MemorySink;
use iaiso_core::{EngineOptions, PressureEngine, StepInput};
use serde::Deserialize;
use serde_json::Value;
use std::path::Path;
use std::sync::Arc;

#[derive(Debug, Deserialize)]
struct EventsStep {
    #[serde(default)]
    tokens: u64,
    #[serde(default)]
    tool_calls: u64,
    #[serde(default)]
    depth: u64,
    #[serde(default)]
    tag: Option<String>,
    #[serde(default)]
    reset: bool,
}

#[derive(Debug, Deserialize)]
struct ExpectedEvent {
    #[serde(default)]
    schema_version: Option<String>,
    #[serde(default)]
    execution_id: Option<String>,
    kind: String,
    #[serde(default)]
    data: Option<Value>,
}

#[derive(Debug, Deserialize)]
struct EventsVector {
    name: String,
    #[serde(default)]
    description: String,
    #[serde(default)]
    config: Value,
    #[serde(default)]
    clock: Vec<f64>,
    #[serde(default)]
    steps: Vec<EventsStep>,
    #[serde(default)]
    reset_after_step: Option<usize>,
    #[serde(default)]
    clock_after_reset: Option<f64>,
    execution_id: String,
    expected_events: Vec<ExpectedEvent>,
}

#[derive(Debug, Deserialize)]
struct EventsFile {
    vectors: Vec<EventsVector>,
}

pub fn run_events(spec_root: &Path) -> Result<Vec<VectorResult>, String> {
    let path = spec_root.join("events").join("vectors.json");
    let data = std::fs::read(&path).map_err(|e| format!("read {}: {}", path.display(), e))?;
    let file: EventsFile =
        serde_json::from_slice(&data).map_err(|e| format!("parse {}: {}", path.display(), e))?;

    let mut results = Vec::with_capacity(file.vectors.len());
    for v in file.vectors {
        results.push(run_one_events(v));
    }
    Ok(results)
}

fn run_one_events(v: EventsVector) -> VectorResult {
    let mut r = VectorResult {
        section: "events".to_string(),
        name: v.name.clone(),
        passed: false,
        message: String::new(),
    };
    let cfg = config_from_value(&v.config);
    let clock = scripted_clock(v.clock.clone());
    let sink = Arc::new(MemorySink::new());

    let engine = match PressureEngine::new(
        cfg,
        EngineOptions {
            execution_id: v.execution_id.clone(),
            audit_sink: sink.clone(),
            clock: clock.clone(),
            timestamp_clock: clock,
        },
    ) {
        Ok(e) => e,
        Err(e) => {
            r.message = format!("engine construction failed: {}", e);
            return r;
        }
    };

    for (i, step) in v.steps.iter().enumerate() {
        if step.reset {
            engine.reset();
        } else {
            engine.step(StepInput {
                tokens: step.tokens,
                tool_calls: step.tool_calls,
                depth: step.depth,
                tag: step.tag.clone(),
            });
        }
        // 1-based: reset_after_step = N triggers after running step N
        if let Some(after) = v.reset_after_step {
            if i + 1 == after {
                engine.reset();
            }
        }
    }

    let got = sink.events();
    if got.len() != v.expected_events.len() {
        r.message = format!(
            "event count: got {}, want {}",
            got.len(),
            v.expected_events.len()
        );
        return r;
    }

    for (i, exp) in v.expected_events.iter().enumerate() {
        let actual = &got[i];
        if let Some(sv) = &exp.schema_version {
            if !sv.is_empty() && actual.schema_version != *sv {
                r.message = format!(
                    "event {} schema_version: got {}, want {}",
                    i, actual.schema_version, sv
                );
                return r;
            }
        }
        if let Some(eid) = &exp.execution_id {
            if !eid.is_empty() && actual.execution_id != *eid {
                r.message = format!(
                    "event {} execution_id: got {}, want {}",
                    i, actual.execution_id, eid
                );
                return r;
            }
        }
        if actual.kind != exp.kind {
            r.message = format!("event {} kind: got {}, want {}", i, actual.kind, exp.kind);
            return r;
        }
        if let Some(want_data) = &exp.data {
            if !data_equal(&actual.data, want_data) {
                r.message = format!(
                    "event {} data mismatch: got {:?}, want {:?}",
                    i, actual.data, want_data
                );
                return r;
            }
        }
    }
    r.passed = true;
    r
}

/// Loose-equality comparison: missing keys in actual that map to nil in
/// expected are treated as equivalent; numbers compared with TOLERANCE.
fn data_equal(actual: &std::collections::BTreeMap<String, Value>, want: &Value) -> bool {
    let want_obj = match want.as_object() {
        Some(o) => o,
        None => return false,
    };
    for (k, w) in want_obj {
        let a = actual.get(k).cloned().unwrap_or(Value::Null);
        if !value_equal(&a, w) {
            return false;
        }
    }
    true
}

fn value_equal(a: &Value, b: &Value) -> bool {
    match (a, b) {
        (Value::Null, Value::Null) => true,
        (Value::Bool(x), Value::Bool(y)) => x == y,
        (Value::String(x), Value::String(y)) => x == y,
        (Value::Number(x), Value::Number(y)) => match (x.as_f64(), y.as_f64()) {
            (Some(xf), Some(yf)) => (xf - yf).abs() <= crate::TOLERANCE,
            _ => x == y,
        },
        (Value::Array(x), Value::Array(y)) => {
            x.len() == y.len() && x.iter().zip(y).all(|(xv, yv)| value_equal(xv, yv))
        }
        (Value::Object(x), Value::Object(y)) => {
            x.len() == y.len()
                && x.iter()
                    .all(|(k, v)| y.get(k).map_or(false, |w| value_equal(v, w)))
        }
        _ => false,
    }
}
